[ESIMD][DOC] Update description of accessor-based memory APIs (#9582)
intel/llvm: sycl/doc/extensions/experimental/sycl_ext_intel_esimd/sycl_ext_intel_esimd.md

Explicit SIMD memory access interface is quite different from the standard SYCL
memory access interface. It supports main SYCL's device memory representations:
- USM pointers
- SYCL accessors
  - 1D global accessors
  - 1D local accessors
  - 2D image accessors

> Only small subset of `sycl::accessor` APIs is supported in ESIMD context:

Reviewer: ```suggestion
Only a small subset of `sycl::accessor` APIs are supported in ESIMD context:
```

- accessor::accessor();

> - accessor::get_pointer(); // Supported only with `-fsycl-esimd-force-stateless-mem` switch.

Reviewer: ```suggestion
- accessor::get_pointer(); // Supported only with the `-fsycl-esimd-force-stateless-mem` switch.
```

> - accessor::operator[]; // Supported only with `-fsycl-esimd-force-stateless-mem` switch.

Reviewer: ```suggestion
- accessor::operator[]; // Supported only with the `-fsycl-esimd-force-stateless-mem` switch.
```

ESIMD provides special APIs to access memory through accessors. Those APIs

> accept accessor object as a base reference to the addressed memory and

Reviewer: ```suggestion
an accept accessor object as a base reference to the addressed memory and
```

Author: you probably mean 'accept an accessor object'. I'll fix it that way.

> additional offsets operand(s) to work with accessed buffer or image.

Reviewer: ```suggestion
additional offset operand(s) to work with accessed buffer or image.
```

> additional offsets operand(s) to work with accessed buffer or image.

Reviewer: ```suggestion
additional offsets operand(s) to work with accessed buffers or images.
```

#### Shared local memory access

The above APIs are available for the global memory as well as shared local
memory (SLM). SLM is allocated in cache memory and thus is much faster.
This memory is shared between work items in a workgroup - basically
it is ESIMD variant of the SYCL `local` memory.

> SLM variants of APIs have 'slm_' prefix in their names,

Reviewer: ```suggestion
SLM variants of APIs have the 'slm_' prefix in their names,
```

e.g. slm_block_load(), or lsc_slm_gather().

SLM memory must be explicitly allocated before it is read or written.

There are 3 different ways of SLM allocation in ESIMD:
* static allocation using slm_init<SLMByteSize>() and slm_init(SpecializationConstSLMByteSize)
* semi-dynamic allocation using slm_allocator<SLMByteSize> class
* SYCL local accessors

##### Static allocation of SLM using slm_init function.
Static allocation is done with help of one of two `slm_init` functions shown
below. The SLM byte size passed to `slm_init` must be either a compilation time
constant or a specialization constant.

```cpp
// Static allocation of `size` bytes of SLM memory.
// Can be used together with slm_allocator() class for
// semi-dynamic SLM allocation of extra SLM.
template <uint32_t size> void slm_init();

// This function is recommended ONLY when 'size' is a specialization constant.

> // It cannot be used together with slm_allocator() class.

Reviewer: ```suggestion
// It cannot be used together with the slm_allocator() class.
```

version of `slm_init` - slm_init(NumBytes);
* `slm_allocator` cannot be used in ESIMD functions called via `invoke_simd()`
or functions not reachable from ESIMD kernel at compilation time.

Example:
```cpp
    // main kernel
    q.parallel_for(Size, [=](id<1> i)[[intel::sycl_explicit_simd]]{
      slm_init<SLMBaseSize>();
      {
        // Use the first SLMBaseSize bytes of SLM here
      }
      foo();
      bar();
      {
        // Continue using the first SLMBaseSize bytes of SLM here.
      }
    }).wait_and_throw();```
    ...

void foo() {
  slm_allocator<SLM_FOO> foo_slm; // allocate extra SLM_FOO bytes of SLM
  unsigned slm_offset = foo_slm.get_offset();
  /* Work with SLM memory: [slm_offset : slm_offset + SLM_FOO - 1] */
}
void bar() {
  slm_allocator<SLM_BAR> bar_slm; // allocate extra SLM_BAR bytes of SLM
  unsigned slm_offset = bar_slm.get_offset();
  /* Work with SLM memory: [slm_offset : slm_offset + SLM_BAR - 1] */
}
```
The actual amount of SLM memory allocated for the kernel above would be:
`SLM_BASE_SIZE + max(SLM_FOO, SLM_BAR)`.

##### Local accessors
Regular SYCL local accessors can be passed from HOST to ESIMD kernel, be used

> in ESIMD kernel and functions called from kernel. It can also be passed from

Reviewer: ```suggestion
in an ESIMD kernel and functions called from kernel. It can also be passed from
```

> SYCL context to ESIMD using `invoke_simd()` API. Similarly to regular SYCL

Reviewer: ```suggestion
SYCL context to ESIMD using `invoke_simd()` API. Similar to regular SYCL
```

> kernels, more than 1 local accessor can be passed to ESIMD kernel.

Reviewer: ```suggestion
kernels, more than 1 local accessor can be passed to an ESIMD kernel.
```

> Please notice that currently the local accessor can be passed through

Reviewer: ```suggestion
Please note that currently the local accessor can be passed through
```

> `invoke_simd()` only as a uniform pointer. Passing local accessor by-value is

Reviewer: ```suggestion
`invoke_simd()` only as a uniform pointer. Passing the local accessor by-value is
```

> not yet supported in GPU driver.

Reviewer: ```suggestion
not yet supported in the GPU driver.
```
